src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -273,14 +273,25 @@ def iter_index_candidate_infos() -> Iterator[IndexCandidateInfo]:
             )
             icans = list(result.iter_applicable())
 
-            # PEP 592: Yanked releases must be ignored unless only yanked
-            # releases can satisfy the version range. So if this is false,
-            # all yanked icans need to be skipped.
+            # PEP 592: Yanked releases are ignored unless the specifier
+            # explicitely pins a version (via '==' or '===') that can be
+            # solely satisfied by a yanked release.
             all_yanked = all(ican.link.is_yanked for ican in icans)
 
+            def is_pinned(specifier: SpecifierSet) -> bool:
+                for sp in specifier:
+                    if sp.operator == "===":
+                        return True
+                    if sp.operator != "==":
+                        continue
+                    if sp.version.endswith(".*"):
+                        continue
+                    return True
+                return False
+
             # PackageFinder returns earlier versions first, so we reverse.
             for ican in reversed(icans):
-                if not all_yanked and ican.link.is_yanked:
+                if (all_yanked and not is_pinned(specifier)) and ican.link.is_yanked:
```

Comment:
Is this condition correct? It would seem that if some of the releases are **not** yanked, then we would want to skip yanked releases. Currently, this turns into condition:
```
if (False and ...) and ican.link.is_yanked:
    ...
```
which means that the yanked releases won't get ignored.
If I understand this correctly, it should instead be:
```suggestion
                if not (all_yanked and is_pinned(specifier)) and ican.link.is_yanked:
```

---

Also I've noticed that `is_pinned()` is called for each ican even though the `specifier` argument is always the same. It's probably a premature optimization but it seems that this could be calculated just once.

Reply:
> Is this condition correct? It would seem that if some of the releases are not yanked, then we would want to skip yanked releases. 

This is actually a good point, I trusted `tests/functional/test_install.py::test_ignore_yanked_file` for this but it turns out there might be a bug/undocumented feature in the code. 

The test is pass, but only because  the `icans` returned by `PackageFinder` looks like this:
```python
ican=<InstallationCandidate('simple', <Version('3.0')>, <Link file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/packages/simple-3.0.tar.gz (from file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/indexes/yanked/simple/index.html)>)>
ican=<InstallationCandidate('simple', <Version('1.0')>, <Link file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/packages/simple-1.0.tar.gz (from file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/indexes/yanked/simple/index.html)>)>
ican=<InstallationCandidate('simple', <Version('2.0')>, <Link file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/packages/simple-2.0.tar.gz (from file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/indexes/yanked/simple/index.html)>)>
```

When reversed later, `simple-2.0` becomes the first package, which passes the in-loop condition as it is not yanked and is therefore installed, making the test pass. 

Altering the index used by the test and un-yanking `simple-3.0` corrects the list order returned by `PackageFinder`:
```python
ican=<InstallationCandidate('simple', <Version('1.0')>, <Link file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/packages/simple-1.0.tar.gz (from file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/indexes/yanked/simple/index.html)>)>
ican=<InstallationCandidate('simple', <Version('2.0')>, <Link file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/packages/simple-2.0.tar.gz (from file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/indexes/yanked/simple/index.html)>)>
ican=<InstallationCandidate('simple', <Version('3.0')>, <Link file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/packages/simple-3.0.tar.gz (from file:///tmp/pytest-of-root/pytest-0/test_ignore_yanked_file0/data/indexes/yanked/simple/index.html)>)>
```

Hence, I am not sure on how/whether to fix this. If we can rely on this returned order, then both the code and the test are valid. Otherwise, we have to think about a new condition for the in-loop continue clause.

> `if not (all_yanked and is_pinned(specifier)) and ican.link.is_yanked:`

This is fine for me as it passes both tests.

> Also I've noticed that `is_pinned()` is called for each ican even though the specifier argument is always the same. It's probably a premature optimization but it seems that this could be calculated just once.

Good idea, I can adapt as proposed.

@uranusjr Should I implement both the proposed changes?

Reply:
Let's fix the `is_pinned()` thing.

The yanked ordering one is more complicated. `PackageFinder` ordering yanked versions last is a designed feature, defined here:

https://github.com/pypa/pip/blob/main/src/pip/_internal/index/package_finder.py#L473-L539

So it's OK to rely on this behaviour. The question is whether it's a good idea to do this when we have another solution. The answer to this is usually no, so let's change it. I actually have a feeling we might change the condition again later here specifically due to the prerelease issue (we'd want `PackageFinder` to return those, but still only allow them if all stable releases are yanked, and I think that's only doable if we rely on ordering). But let's worry about that later.

Reply:
Agreed, done.